src/main/java/org/radarbase/output/worker/KafkaCleanerWorker.kt

```diff
@@ -0,0 +1,103 @@
+package org.radarbase.output.worker
+
+import org.apache.avro.generic.GenericRecord
+import org.radarbase.output.FileStoreFactory
+import org.radarbase.output.accounting.Accountant
+import org.radarbase.output.source.SourceStorage
+import org.radarbase.output.source.TopicFile
+import org.radarbase.output.util.Timer.time
+import org.radarbase.output.worker.RestructureWorker.Companion.extractRecords
+import org.slf4j.LoggerFactory
+import java.io.Closeable
+import java.io.IOException
+import java.nio.file.Path
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+import java.util.concurrent.atomic.AtomicBoolean
+
+class KafkaCleanerWorker(
+        private val fileStoreFactory: FileStoreFactory,
+        private val accountant: Accountant,
+        private val sourceStorage: SourceStorage,
+        private val closed: AtomicBoolean
+) : Closeable {
+    private val maxFilesPerTopic: Int = fileStoreFactory.config.worker.maxFilesPerTopic ?: Int.MAX_VALUE
+    private val pathFactory = fileStoreFactory.pathFactory
+    private val reader = sourceStorage.createReader()
+
+    private val deleteThreshold: Instant? = Instant.now().apply { minus(fileStoreFactory.config.cleaner.age.toLong(), ChronoUnit.DAYS) }
+
+    private val cacheStore = ContainsFileCacheStore(fileStoreFactory)
+
+    fun deleteOldFiles(
+            topic: String,
+            topicPath: Path
+    ): Int {
+        val offsets = accountant.offsets.copyForTopic(topic)
+        return sourceStorage.walker.walkRecords(topic, topicPath)
+                .filter { f -> f.lastModified.isBefore(deleteThreshold) &&
+                        // ensure that there is a file with a larger offset also
+                        // processed, so the largest offset is never removed.
+                        offsets.contains(f.range.copy(range = f.range.range.copy(to = f.range.range.to + 1))) }
```

Comment:
There may not be an offset exactly +1 but something larger. Does this take that in account too?

Reply:
Nope. It will not try to delete the file then. It should be a relatively rare occurrence, and the penalty for not deleting a file is rather low: a single file does not take much space and it can still manually be removed.